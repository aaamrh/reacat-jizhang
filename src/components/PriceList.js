import React, { Component } from 'react';
import PropTypes from 'prop-types';


const PriceList = ({ items, onModifyItem, onDeleteItem }) => {

  return  (
    <ul className='list'>
      {
        items.map( item => {
          return <li key={item.id}>
            <span> { item.category.name} </span>
            <span> { item.title } </span>
            <span> 
              { item.category.type === 'income' ? '+' : '-' }
              { item.price } 
              元 
            </span>
            <span> { item.date } </span>
            <button onClick={ ()=>{ onModifyItem(item) } }> 编辑 </button>
            <button onClick={ ()=>{ onDeleteItem(item) } }> 删除 </button>
          </li>
        } )
      }
    </ul>
  )
}

PriceList.propTypes = {
  items: PropTypes.array.isRequired,
  onModifyItem: PropTypes.func.isRequired,
  onDeleteItem: PropTypes.func.isRequired
}

PriceList.defaultProps = {
  items: PropTypes.array.isRequired,
  onModifyItem: ()=>{},
  onDeleteItem: ()=>{}
}
export default PriceList