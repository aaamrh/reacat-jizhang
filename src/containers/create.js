import React, { Component } from 'react';
import {withRouter} from 'react-router-dom'
import CategorySelect from '../components/CategorySelect';
import PriceForm from '../components/PriceForm';
import { testCategories } from '../testData'
import { Tabs, Tab } from '../components/Tabs';
import { TYPE_INCOME, TYPE_OUTCOME } from '../utils'
import { AppContext } from '../App'


const tabsText = [ TYPE_OUTCOME, TYPE_INCOME ]

class Create extends Component {
  constructor(props){
    super(props)

    this.state = {
      selectedTab: TYPE_OUTCOME,
      selectedCategory: null,
    }
  }

  cancelSubmit = () => {
    this.props.history.push('/')
  }

  selectCategory = (category) => {
    this.setState({
      selectedCategory: category
    })
  }

  tabChange = (index) => {
    this.setState({
      selectedTab: tabsText[index]
    })
  }

  render() {
    // const filterCategories = testCategories.filter( category => category.type===TYPE_OUTCOME )

    return (
      <AppContext.Consumer>
        { ( {state, actions} )=>{
          const { id } = this.props.match.params;
          const { items, categories } = state;
          const selectedTab = (id && items[id]) ? categories[ items[id].cid ].type : this.state.selectedTab;
          const selectedCategory = (id && items[id]) ? categories[ items[id].cid ] : this.state.selectedCategory

          const editItem = (id && items[id] ) ? items[id] : {}

          const filterCategories = Object.keys(categories)
          .filter( id=>categories[id].type === selectedTab)
          .map(id => categories[id])

          const submitForm = (data, isEditMode) => {
            if ( !isEditMode ){
              actions.createItem(data, selectedCategory.id)
            }else{
        
            }
            this.props.history.push('/')
          }
          
          return (
            
            <div className="create-page py-3 px-3 rounded mt-3" style={{ background: '#fff' }}>
              <Tabs activeIndex={0} onTabChange={ this.tabChange }>
                <Tab>支出</Tab>
                <Tab>收入</Tab>
              </Tabs>

              <CategorySelect categories={filterCategories} 
                onSelectCategory={ this.selectCategory }
                selectedCategory = { selectedCategory }
              />

              <PriceForm 
                onFormSubmit={ submitForm }
                onCancelSubmit={ this.cancelSubmit }
                item={editItem}
              />
            </div>
          )
        } }
      </AppContext.Consumer>

      // <div>
      //   create.js { this.props.match.params.id }
      //   {/* <CategorySelect categories={filterCategories} 
      //     onSelectCategory={this.selectCategory}
      //     selectedCategory={selectedCategory}
      //   /> */}
      //   <PriceForm 
      //     onFormSubmit={this.submitForm}
      //     onCancelSubmit={this.cancelSubmit}
      //     item={{}}
      //   />
      // </div>
    );
  }
}

export default withRouter ( Create );