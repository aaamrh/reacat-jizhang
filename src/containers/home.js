import React, { Component } from 'react';

import { LIST_VIEW, CHART_VIEW, parseToYearAndMonth } from '../utils'
import logo from '../logo.svg';

import PriceList from '../components/PriceList'
import ViewTab from '../components/ViewTab';
import TotalPrice from '../components/TotalPrice';
import MonthPicker from '../components/MonthPicker';
import CreateBtn from '../components/CreateBtn';


export const categories = {
  '0': {
    'id': 0,
    'name': '工资',
    'type': 'income'   
  },
  '1':{
    'id': 1,
    'name': '旅行',
    'type': 'outcome'   
  },
  '2':{
    'id': 2,
    'name': '零食',
    'type': 'outcome'   
  }

}

export const items = [
  {
    'id': 0,
    'title': '工资',
    'price': 5000,
    'date': '2020-12-10',
    'cid': 0
  },
  {
    'id': 1,
    'title': '去云南旅游',
    'price': 25,
    'date': '2021-1-11',
    'cid':1
  },
  {
    'id': 2,
    'title': '车票',
    'price': 115,
    'date': '2021-1-15',
    'cid':2
  }
]

const newItem = {
  'id': 4,
  'title': '新纪录',
  'price': 280,
  'date': '2021-1-18',
  'cid':1
}

class Home extends Component {
  constructor(props){
    super(props)
    this.state = {
      items,
      currentDate: parseToYearAndMonth(),
      tabView: LIST_VIEW
    }
  }


  changeView = (view) => {
    this.setState({
      tabView: view
    })
  }

  changeDate = (year, month) => {
    this.setState({
      currentDate: {year, month}
    })
  }

  createItem = () => {
    this.setState({
      items: [newItem, ...this.state.items]
    })
  }

  modifyItem = (modifiedItem) => {
    const modifiedItems = this.state.items.map( item => {
      if( item.id === modifiedItem.id ){
        return { ...item, title: '被修改' }
      }
      return item
    } )

    this.setState({
      items: modifiedItems
    })
  }
   
  deleteItem = (deletedItem) => {
    const filteredItems = this.state.items.filter( item => item.id !== deletedItem.id )

    this.setState({
      items: filteredItems
    })
  }

  render() {
    const {items, currentDate, tabView} = this.state;

    let totalIncome = 0, totalOutcome = 0;

    const itemsWithCategory = items.map( item => {
      item.category = categories[ item.cid ]
      return item
    } ).filter( item => item.date.includes(`${currentDate.year}-${currentDate.month}`) )

    items.forEach( item => {
      if( item.category.type === 'outcome' ){
        totalOutcome += item.price
      }else{
        totalIncome += item.price
      }
    } )

    return (
      <>
         <header className="App-header">
          <img src={logo} className="App-logo" alt="logo" />
        </header>

        <MonthPicker 
          year={currentDate.year}
          month={currentDate.month}
          onChange = { this.changeDate }
        />

        <TotalPrice 
          income = {totalIncome}
          outcome = {totalOutcome}
        />

        <CreateBtn onClick={ this.createItem } />

        <ViewTab 
          activeTab = { tabView }
          onTabChange = { this.changeView }
        />
        {
          tabView === LIST_VIEW && 
          <PriceList items={itemsWithCategory} 
            onModifyItem = {this.modifyItem}
            onDeleteItem = {this.deleteItem}
          />
        }
        {
          tabView === CHART_VIEW && 
          <h1>图表</h1>
        }
      </>
    );
  }
}

export default Home;