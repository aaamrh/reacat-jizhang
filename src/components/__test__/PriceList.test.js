import { shallow } from 'enzyme'  
import PriceList from '../PriceList'
import { categories, items } from '../../containers/home'

const itemsWithCategory = items.map( item => {
  item.category = categories[ item.cid ]
  return item
} )

const props = {
  items: itemsWithCategory,
  onModifyItem: jest.fn(),
  onDeleteItem: jest.fn()
}

let wrapper;
describe('test PriceList component', ()=>{
  beforeEach( ()=>{
    wrapper = shallow(<PriceList {...props} />)
  } ) 
  // wrapper = shallow(<PriceList {...props} />)

  it('match snapshot', ()=>{
    expect(wrapper).toMatchSnapshot()
  })


  it('render correct price items length', ()=>{
    expect(wrapper.find('.list li').length).toEqual(itemsWithCategory.length)
  })

  it('trigger the correct func callbacks', ()=>{
    const firstItem = wrapper.find('.list li').first()

    firstItem.find('button').first().simulate('click')
    expect( props.onModifyItem ).toHaveBeenCalledWith(itemsWithCategory[0] )
    
    firstItem.find('button').last().simulate('click') 
    expect( props.onModifyItem ).toHaveBeenCalledWith(itemsWithCategory[0] )
  })
})


