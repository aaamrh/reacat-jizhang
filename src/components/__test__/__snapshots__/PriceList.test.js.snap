// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test PriceList component match snapshot 1`] = `ShallowWrapper {}`;
